lib/marked.js

```diff
@@ -505,7 +505,7 @@ inline._punctuation = '!"#$%&\'()*+,\\-./:;<=>?@\\[^_{|}~';
 inline.em = edit(inline.em).replace(/punctuation/g, inline._punctuation).getRegex();
 
 inline._inside = /(?:\[[^\]]*\]|[^\[\]]|\](?=[^\[]*\]))*/;
-inline._href = /(?:[^()]|\([^()]*\))*/;
+inline._href = /(?:[^()]*|\([^()]*\)*)*/;
```

Comment:
The asterisk in the first part of this doesn't do anything since it basically turns that part of the regex into `(?:[^()]*)*` which causes a lot of problems in a regex.

You can simplify the `[^()]` pattern to just `a` to see the problem better. If you do that here, you end up with the pattern `(?:a*)*`. The problem with that pattern as well as the original is that it can cause some performance issues because it can group the letters in a string in many different ways when searching for a match. For example, if you have a string like `aaaa` and you try to find a match for either of those patterns, they will both check for matches by grouping the letters like `(aaaa)`, `(aaa)(a)`, `(aa)(aa)`, `(aa)(a)(a)`, and so on which causes the code to check an exponentially increasing number of different cases. Without the extra asterisk, those patterns would look like `(?:[^()])*` and `(?:a)*`, and they only groups the letters one way: `(a)(a)(a)(a)`.

The second part of this also has the same issue because you've made the final bracket optional, so you also end up with basically `(?:\(a*)*` which exhibits the same issue as long as there's an opening bracket somewhere in the string.

It's actually impossible to use regular expressions to parse arbitrarily nested brackets, so I don't have any particularly good advice for solving this for any number of nested brackets outside of just implementing proper URL parsing. If you want to add support for brackets nested two deep (like `(())` as in the ticket), you could look at the `gfm.url` regex below since that should support that. It is a lot more complex as you can see though.

Reply:
Thanks @hmhealey.  Regex's have not been my strong suite, but very open to improving upon that.  I just tried using the gfm.url regex though had some test failures.  I will look more into this.

I also tried some of the regex's from the main marked repo, though they didn't pass tests either.  The latest 6.2 release does parse the double () correctly, so I am hopeful a solution can be found for the ticket.

Reply:
Yeah, I saw theirs. I'm surprised it's more complicated than ours, but that's probably to support things like nested brackets.

Also, yeah, regexes are hard. I've mostly learned about them the hard way by breaking Mattermost by adding things just like that. 😛